File: algorithms_i/hw_1/count_inversions.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

long long count_split_inv(long array[], long lower, long mid, long upper){

    long left_pos = lower;
    long left_end = mid;
    long right_pos = mid+1;
    long right_end = upper;

    // No need to add one because the first index is 0
    long tmp_values[upper - lower];
    long long nr_inversions = 0;
    long i = 0;

    long k;

    while ((left_pos <= left_end) && (right_pos <= right_end)){
        if (array[left_pos] <= array[right_pos]){
            tmp_values[i] = array[left_pos];;
            i++;
            left_pos++;
        }
        else {
            nr_inversions = nr_inversions + (left_end - left_pos + 1);
            tmp_values[i] = array[right_pos];
            i++;
            right_pos++;
        }
    }

    // When the previous while finishes we have to parse one
    // of the sides until the end
    if (left_pos > left_end){
        for (k = right_pos; k <= right_end; k++, i++){
            tmp_values[i] = array[k];
        }
    }
    else {
        for (k = left_pos; k <= left_end; k++, i++){
            tmp_values[i] = array[k];
        }
    }

    // Merge the resulting array into the array
    for (k = lower; k <= upper; k++){
        array[k] = tmp_values[k - lower];
    }

    return nr_inversions;

}

long long count(long array[], long lower, long upper){

    long mid = lower + (upper - lower) / 2;
    if (lower >= upper){
        return 0;
    }
    else {
        long nr_left_inversions = count(array, lower, mid);
        long nr_right_inversions = count(array, mid + 1, upper);
        long nr_split_inversions = count_split_inv(array, lower, mid, upper);
        return nr_left_inversions + nr_right_inversions + nr_split_inversions;
    }
}

int main(){

    std::ifstream file("IntegerArray.txt");
    std::string line;

    // Get the number of lines in the file
    int nr_lines = 0;

    while (std::getline(file, line)){
        nr_lines ++;
    }

    std::cout << "Number of lines: " << nr_lines << std::endl;

    // Read once again to fill the array
    file.clear();
    file.seekg(0);

    long values[nr_lines];
    long value;
    long i = 0;

    while (std::getline(file, line)){
        value = atoi(line.c_str());
        values[i] = value;
        i++;
    }

    std::cout << "Number of parsed values: " << i << std::endl;

    // Calculate the number of inversions
    long nr_inversions = count(values, 0, nr_lines-1);
    std::cout << "Number of inversions: " << nr_inversions << std::endl;

    return 0;

}

File: algorithms_i/hw_1/count_inversions.py
def sort_count(array):
    if len(array) == 1:
        return array, 0
    else:
        mid = len(array)//2
        left_sorted, left_inversions = sort_count(array[:mid])
        right_sorted, right_inversions = sort_count(array[mid:])
        split_sorted, split_inversions = count_split_inv(left_sorted, right_sorted)
        return split_sorted, left_inversions+right_inversions+split_inversions

def count_split_inv(left_array, right_array):
    i = 0
    j = 0
    nr_inversions = 0
    merged_array = []

    while (i < len(left_array) and j < len(right_array)):
        if left_array[i] < right_array[j]:
            merged_array.append(left_array[i])
            i += 1
        else:
            merged_array.append(right_array[j])
            nr_inversions += len(left_array[i:])
            j += 1

    merged_array.extend(left_array[i:])
    merged_array.extend(right_array[j:])
    return merged_array, nr_inversions

def main():
    with open('IntegerArray.txt') as f:
        array = [int(x) for x in f]
    print("Number of inversions: ", sort_count(array)[1])

if __name__ == "__main__":
    main()
